File: 39-combination-sum/39-combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    let ans = []
    let n = candidates.length
    let ds = []
    const rec = (i, t) => {
        if(t==0){
            ans.push(ds.slice())
        }
        if(i==n) return
        if(t<=0) return
        
        if(candidates[i]<=t){
            ds.push(candidates[i])
            rec(i, t-candidates[i])
            ds.pop()
        }
        rec(i+1, t)
    }
    
    rec(0, target)
    
    return ans
};

File: 39-combination-sum/39-combination-sum.py
class Solution:
    def combinationSum(self, A, tar) -> List[List[int]]:
        
        ans = []
        def rec(i, ds, rem):
            if rem==0:
                ans.append(ds[:])
            
            if rem<=0: return
            if i>=len(A): return
            
            if rem>=A[i]:
                ds.append(A[i])
                rem-=A[i]
                rec(i, ds, rem)
                rem+=A[i]
                ds.pop(len(ds)-1)
                
            rec(i+1, ds, rem)
            
        
        rec(0, [], tar)
        
        return ans